Privacy experience form

allisonking opened the pull request:
Closes https://github.com/ethyca/fides/issues/3126

### Code Changes

* [x] Add new `PrivacyExperienceForm.tsx`
  * [x] This one has a few subcomponents which control the various form rendering logic from privacy experiences
  * [x] Other than that, it's pretty similar to `PrivacyNoticeForm.tsx`
* [x] Cypress tests, including a handy function to easily test different experiences backed by varying notices

### Steps to Confirm

* Same setup as described in https://github.com/ethyca/fides/pull/3186
* Click into one of the experiences
* You should be able to fill out the form and persist the data 

### Pre-Merge Checklist

* [x] All CI Pipelines Succeeded
* Documentation:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
* [x] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [ ] Update `CHANGELOG.md`
* [ ] For API changes, the [Postman collection](https://github.com/ethyca/fides/blob/main/docs/fides/docs/development/postman/Fides.postman_collection.json) has been updated

### Description Of Changes

https://www.loom.com/share/556bc39d3f6b4f7ca50a4c2cef14929a


https://user-images.githubusercontent.com/24641006/236256514-8be4a0c7-e771-4719-bb93-9ce28ef5571b.mov

TheAndrewJackson reviewed (approved):
I played around with all the example experiences the original PR had and was able to edit everything as expected. I like the implementation for dynamically showing different parts of the form.


I think it was a good choice to not attempt a dynamic validation schema like I did with the custom fields 😅 At least not for the first pass. I wish that formik supported that pattern better.


:shipit:

allisonking commented:
> I played around with all the example experiences the original PR had and was able to edit everything as expected. I like the implementation for dynamically showing different parts of the form.
> 
> I think it was a good choice to not attempt a dynamic validation schema like I did with the custom fields 😅 At least not for the first pass. I wish that formik supported that pattern better.
> 
> :shipit:

haha yeah, I thought I could get away with no ValidationSchema, but it turns out we probably do want some validation https://github.com/ethyca/fides/issues/3226 so I might have to restructure how this works

allisonking commented:
I refactored the form logic a bit in order to accommodate required fields (somewhat dynamic validation schemas—not to the same extent you had to do @TheAndrewJackson in custom fields, since they don't change on user input, just depending on what data is received). I added a hook `useExperienceForm` to encapsulate the various rule logic, so that all rules can hang out in one place, which hopefully makes adding/editing/removing rules easier

TheAndrewJackson reviewed (approved):
Solid refactor. The new hook cleans the logic up a lot. It's helpful that the validation logic doesn't need to change based on a drop down option! It's still dynamic but not as dynamic as the custom fields form which is nice.

allisonking commented:
I had a suspicion my hook was being called more than I wanted it to be and it turned out to be true! so I refactored one more time 😅 this time rules are passed around as props, but I think it works pretty well since the same set of rules get passed everywhere
